hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterWebHdfsMethods.java

```diff
@@ -454,19 +454,12 @@ private URI redirectURI(final Router router, final UserGroupInformation ugi,
   private DatanodeInfo chooseDatanode(final Router router,
       final String path, final HttpOpParam.Op op, final long openOffset,
       final String excludeDatanodes) throws IOException {
-    // We need to get the DNs as a privileged user
     final RouterRpcServer rpcServer = getRPCServer(router);
-    UserGroupInformation loginUser = UserGroupInformation.getLoginUser();
-    RouterRpcServer.setCurrentUser(loginUser);
-
     DatanodeInfo[] dns = null;
     try {
-      dns = rpcServer.getDatanodeReport(DatanodeReportType.LIVE);
+      dns = rpcServer.getCachedDatanodeReport(DatanodeReportType.LIVE);
```

Comment:
Question from me and not quite related to this JIRA: seems this is getting DNs for all the clusters while in the `CREATE` case we should only choose DN from a particular sub-cluster. Where is this logic implemented?

Reply:
You are correct. I found this issue also. We could be redirected to a datanode from a different subcluster where the path is acutually mounted, but the file would be written to the correct subcluster. We may need to open a new JIRA for this.

Reply:
Thanks. I filed https://issues.apache.org/jira/browse/HDFS-15423 to track this.